akka-javasdk/src/main/scala/akka/javasdk/impl/serialization/JsonSerializer.scala

```diff
@@ -0,0 +1,265 @@
+import com.fasterxml.jackson.core.JsonProcessingException
+
+object JsonSerializer {
+  val JsonContentTypePrefix: String = "json.akka.io/"
+  private val KalixJsonContentTypePrefix: String = "json.kalix.io/"
+
+  final case class TypeHint(currenTypeHintWithVersion: String, allTypeHints: List[String])
+
+}
+
+class JsonSerializer {
+  import JsonSerializer._
+
+  private val typeHints: ConcurrentMap[Class[_], TypeHint] = new ConcurrentHashMap()
+  val reversedTypeHints: ConcurrentMap[String, Class[_]] = new ConcurrentHashMap()
+
+  override def toString: String = s"JsonSerializer: ${typeHints.keySet().size()} registered types"
+
+  private val objectMapper = JsonSupport.getObjectMapper
+
+  def toBytes(value: Any): BytesPayload = {
+    if (value == null) throw NullSerializationException
+    val typeHint = lookupTypeHintWithVersion(value)
+    val byteArray = objectMapper.writerFor(value.getClass).writeValueAsBytes(value)
+    new BytesPayload(bytes = ByteString.fromArrayUnsafe(byteArray), contentType = JsonContentTypePrefix + typeHint)
+  }
+
+  def fromBytes[T](expectedType: Class[T], bytesPayload: BytesPayload): T = {
+    validateIsJson(bytesPayload)
+
+    try {
+      val migrationAnnotation = expectedType.getAnnotation(classOf[Migration])
+      if (migrationAnnotation != null) {
+        val migration = migrationAnnotation
+          .value()
+          .getConstructor()
+          .newInstance()
+        val fromVersion = parseVersion(bytesPayload.contentType)
+        val currentVersion = migration.currentVersion()
+        val supportedForwardVersion = migration.supportedForwardVersion
+        if (fromVersion < currentVersion) {
+          migrate(expectedType, bytesPayload.bytes, fromVersion, migration);
+        } else if (fromVersion == currentVersion) {
+          parseBytes(expectedType, bytesPayload.bytes)
+        } else if (fromVersion <= supportedForwardVersion) {
+          migrate(expectedType, bytesPayload.bytes, fromVersion, migration)
+        } else {
+          throw new IllegalStateException(
+            s"Migration version [$supportedForwardVersion] is " +
+            "behind version [$fromVersion] of deserialized type [${expectedType.getName}]")
+        }
+      } else {
+        parseBytes(expectedType, bytesPayload.bytes)
+      }
+    } catch {
+      case e: JsonProcessingException =>
+        throw jsonProcessingException(expectedType, bytesPayload.contentType, e)
+      case e @ (_: IOException | _: NoSuchMethodException | _: InstantiationException | _: IllegalAccessException |
+          _: InvocationTargetException) =>
+        throw genericDecodeException(expectedType, bytesPayload.contentType, e)
+    }
+  }
+
+  /**
+   * Parse the bytes to object of type corresponding to the type name in the `bytesPayload.contentType`. Requires that
+   * the types are known by first `registerTypeHints` or calling `contentTypeFor` or `toBytes`.
+   */
+  def fromBytes(bytesPayload: BytesPayload): AnyRef = {
```

Comment:
This was the case where the StrictJsonMessageCodec was used. The event types are registered up front and can then be deserialized without the `expectedType: Class[T]`